File: app/src/main/java/com/softaai/designpatterns/java/creational/prototype/Shape.java
package com.softaai.designpatterns.java.creational.prototype;

/**
 * Created by Amol Pawar on 01-04-2019.
 * softAai Apps
 */
public abstract class Shape implements Cloneable {
    private String id;
    protected String type;
    abstract void draw();

    public String getType(){
        return type;
    }

    public String getId(){
        return id;
    }

    public void setId(String id){
        this.id = id;
    }

    public Object clone() {
        Object clone = null;
        try {
            clone = super.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return clone;
    }
}

File: app/src/main/java/com/softaai/designpatterns/kotlin/creational/prototype/Shape.kt
package com.softaai.designpatterns.kotlin.creational.prototype


/**
 * Created by Amol Pawar on 01-04-2019.
 * softAai Apps
 */
abstract class Shape : Cloneable {
    private lateinit var id: String
    protected lateinit var shapeType:String
    abstract fun draw() : Unit

    public fun getType():String{
        return this.shapeType
    }


    public fun getId() : String{
        return id;
    }

    public fun setId(id : String){
        this.id = id
    }

    override public fun clone(): Any {
        return super.clone()
    }


}